ui/src/adapters/parsers/index.ts

```diff
@@ -0,0 +1,34 @@
+import { ZodSchema, ZodTypeDef } from "zod";
+
+// The following was implemented due to a perceived limitation of Zod:
+// https://github.com/colinhacks/zod/issues/652
+
+// Asserts whether 2 types are exactly the same.
+type Exact<T, U> = [T, U] extends [U, T] ? true : false;
```

Comment:
I'm ok leaving this here, but be aware that this is a generic utility type, unrelated to the adapters and the parsers, that was better located in `utils/types.ts`.

Reply:
But it's only used here. It's directly related to and a part of `getStrictParser`.

Reply:
Indeed, it's only used here, but this is a coincidence.